[MemCpyOpt] allow more memcpy-to-memcpy optimization

vtjnash opened the pull request:
Allow the memcpy-to-memcpy optimization even when the sizes are not identical. For example, it might have been generated as a small slice of a larger struct (currently only for zero offset however), or might be only storing to part of an oversized alloca.

With edits / rewrites by Claude

nikic reviewed (changes requested):
I'm not interested in supporting dynamic allocas in this pass. Can you please limit this to allocas with known size? That should simplify your code a lot.

vtjnash commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1532,21 +1559,43 @@ bool MemCpyOptPass::performStackMoveOptzn(Instruction *Load, Instruction *Store,
+    }
+
+  // Check if it will be legal to combine allocas without breaking dominator.
+  // TODO: Try to hoist the arguments (recursively) instead of giving up
+  // immediately.
```
It appears that request would eliminate this TODO comment and replace isUsedWithInAlloca with isStaticAlloca, but the logic is the same to do the relevant size computations whether at runtime or compile time, where most of the added code is present

nikic replied:
The size domination checks here will go away, and all the dynamic size calculation will go away (you just need a plain max between the type sizes -- or pick the larger alloca, if you want to preserve the type instead of canonicalizing).

vtjnash replied:
Which size domination check goes away?

Writing `llvm.umax()` instrinsic vs `llvm::max()` call seem essentially the same amount of code here, but slightly more annoying to have to convert from an LLVM Value to a compile-time value and back again, when LLVM's already has APIs to do this directly on LLVM Values (InstSimplifyFolder) just as easily

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1508,21 +1532,33 @@ bool MemCpyOptPass::performStackMoveOptzn(Instruction *Load, Instruction *Store,
-  std::optional<TypeSize> SrcSize = SrcAlloca->getAllocationSize(DL);
-  if (!SrcSize || Size != *SrcSize) {
-    LLVM_DEBUG(dbgs() << "Stack Move: Source alloca size mismatch\n");
+  if (SrcAlloca->isUsedWithInAlloca() || DestAlloca->isUsedWithInAlloca())
```
Can we keep this as isStaticAlloca() please?

nikic commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1508,21 +1532,33 @@ bool MemCpyOptPass::performStackMoveOptzn(Instruction *Load, Instruction *Store,
+  // Check that copy is full with dest size, either because it wrote every byte,
+  // or it was fresh.
+  if (Size != *DestSize)
+    if (!allOverreadUndefContents(MSSA, Store, BAA)) {
```
So I'm trying to reason about this and I'm a bit unclear on why we need this allOverreadUndefContents check at all for the case where Size != DestSize.

One of the pre-conditions for the transform (see DestModRefCallback) is that there is no ModRef on Dest prior to the copy operation, so doesn't this mean that this "allOverreadUndefContents" condition should *always* hold?

And as a result we should be able to always do this transform, regardless of how large the copy actually is?

vtjnash replied:
> One of the pre-conditions for the transform (see DestModRefCallback) is that there is no ModRef on Dest prior to the copy operation, so doesn't this mean that this "allOverreadUndefContents" condition should always hold?

Note that `DestModRefCallback` looks like it doesn't technically exactly check that, but rather checks that there is no prior ModRef on Dest *except for this copy operation*. The difference between those could be important if the operation is in a loop and the size is less than the full dest and either the size or the offset is dynamic. The call to allOverreadUndefContents prevents that causing incorrect results in this PR, though an equivalent check could perhaps be integrated directly into DestModRefCallback instead

A big part of my goal with this PR was to figure out ways to reduce the random differences between the various passes in this file (e.g. none of the other passes arbitrarily bail out because of `!isStaticAlloca`–that is already covered by the no-capture information). The use of a custom `DestModRefCallback` here and exact size checks vs. MemorySSA + overreadUndefContents everywhere else seemed like a smell.

It seems like the `DestModRefCallback` design might be a superior design to the MemorySSA checks used elsewhere in this file? In particular, the MemorySSA seems to be very expensive to walk and bails too easily because it has to check every Mod statement (up to an arbitrary cutoff), instead of analyzing only those statements relevant to the alloca, but analyzing them in full. Would you consider reviewing a total replacement for the MemorySSA design in this file with an analysis instead based on this seemingly more robust and efficient DestModRefCallback design? It'll take some time, but I can add it to my backlog of improvements to attempt, if so.

vtjnash replied:
> Just improving the existing DestModRefCallback code to account for the "partial store inside loop" case would be an option

I tried to do this in the latest commit(*) (if it fails all the conditionals such that the check is needed). I think that might really be all that is needed?

(*) Main relevant code is adding:
```cpp
BasicBlock *BB = Store->getParent();
ReachabilityWorklist.append(succ_begin(BB), succ_end(BB));
```

nikic reviewed (commented):
Sorry for the delay

vtjnash commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -1640,6 +1636,46 @@ bool MemCpyOptPass::performStackMoveOptzn(Instruction *Load, Instruction *Store,
+    }
+    // If offset and size are constant, then this overwrote the same bytes every
+    // time anyways.
+    if (!Ptr->getPointerOffsetFrom(DestAlloca, DL)) {
```
Yeah, that should be safe without this, but it feels like a big missing feature for this pass now that you pointed it out. I made a commit to add that optimization now, but need to spend more time write a test and extract that into a separate PR on another day

vtjnash replied:
Support for offsets is now added in https://github.com/llvm/llvm-project/pull/176436, with this PR now rebased on top of that

vtjnash committed:
less conservative allOverreadUndefContents check

vtjnash committed:
less conservative allOverreadUndefContents check

vtjnash committed:
[MemCpyOpt] allow more memcpy-to-memcpy optimization

Allow the memcpy-to-memcpy optimization even when the dest size is not
full with the copy. For example, it might be only storing to part of an
oversized alloca.

vtjnash committed:
[MemCpyOpt] allow more memcpy-to-memcpy optimization

Allow the memcpy-to-memcpy optimization even when the dest size is not
full with the copy. For example, it might be only storing to part of an
oversized alloca.

vtjnash committed:
cleanups

vtjnash committed:
cleanups
